compiler/rustc_lint/src/non_ascii_idents.rs

```diff
@@ -190,17 +191,104 @@ impl EarlyLintPass for NonAsciiIdents {
             if check_uncommon_codepoints
                 && !symbol_str.chars().all(GeneralSecurityProfile::identifier_allowed)
             {
-                let codepoints: Vec<_> = symbol_str
-                    .chars()
-                    .filter(|c| !GeneralSecurityProfile::identifier_allowed(*c))
-                    .collect();
-                let codepoints_len = codepoints.len();
-
-                cx.emit_span_lint(
-                    UNCOMMON_CODEPOINTS,
-                    sp,
-                    IdentifierUncommonCodepoints { codepoints, codepoints_len },
-                );
+                let mut chars = symbol_str.chars().collect::<Vec<_>>();
+
+                let exclusion = chars
```

Comment:
This might a bit too “clever” (i.e., unreadable lol) but instead of duplicating this 5 times, you could consolidate the first 4 searches by doing:

```rs
for (id_ty, id_ty_descr) in [
    (IdentifierType::Exclusion, "Exclusion"),
    (IdentifierType::Technical, "Technical"),
    (IdentifierType::Limited_Use, "Limited_Use"),
    (IdentifierType::Not_NFKC, "Not_NFKC"),
] {
    let codepoints: Vec<_> = chars.extract_if(|(_, ty)| ty == id_ty).collect();
    if codepoints.is_empty() {
        continue;
    }
    cx.emit_span_lint(
        UNCOMMON_CODEPOINTS,
        sp,
        IdentifierUncommonCodepoints {
            codepoints_len: codepoints.len(),
            codepoints,
            identifier_type: id_ty_descr,
        },
    );
}
```